Bullet collision dispatch from RBT

amcastro-tri opened the pull request:
This PR builds on top of #2521 and #2543 to resolve #2004. 
It implements `RBT::ComputeMaximumDepthCollisionPoints` which dispatches collisions using [`Model::ComputeMaximumDepthCollisionPoints`](https://github.com/RobotLocomotion/drake/blob/master/drake/systems/plants/collision/Model.h#L84) and transforms collision points to bodies' frames as needed by [`RigidBodySystem::dynamics` here](https://github.com/RobotLocomotion/drake/blob/master/drake/systems/plants/RigidBodySystem.cpp#L143).

A Google test is implemented to test this capability.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/2596)
<!-- Reviewable:end -->

amcastro-tri commented:
+@sherm1 for platform review.

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-:-KKiYW5eB9cR2oGIh9KV:960604000)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Initial comments. More to come.

---

Reviewed 1 of 6 files at r1.
Review status: 1 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.h, line 671 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKimIx5-syJ7fwUTAYV:-KKimIx5-syJ7fwUTAYW:448263973) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L671)):*
> ```objc
> 
>    @param cache[in] a KinematicsCache constructed by RigidBodyTree::doKinematics
>    given \f$ q \f$ and \f$ v \f$.
> ```

Although this will look lovely in doxygen's documentation, it is hard to read in the header file. It is usually best to compromise so that the text is readable either way, since developers generally only see it in the header. Here I suggest:
```
given `q` and `v`.
```
which looks reasonably good in both places and conveys the idea just as well.

---

*[drake/systems/plants/RigidBodyTree.h, line 677 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKin65RUoZi6sTBzlbC:-KKin65RUoZi6sTBzlbD:-278405711) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L677)):*
> ```objc
> 
>    @param phi[out] A vector containing the closest distance between
>    CollisionElement's that are in contact.
> ```

Actually this method doesn't return which collision elements are in contact. Should it? That information would be very difficult to extract from just the body and contact point location. I noticed that your test case actually used the collision elements but that only worked because you knew there was only one on the body.

---

*[drake/systems/plants/RigidBodyTree.h, line 682 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKiqT6fHcsSGzqONoh9:-KKiqT6fHcsSGzqONohA:1703850767) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L682)):*
> ```objc
>    normal on body A equals the normal on body B but points in the opposite
>    direction, the normal is only reported for body B. This is true only in the
>    world's frame and therefore the normal vector is reported in this frame.
> ```

Word this more precisely. The normal vectors are physical quantities independent of the expressed-in frame; they are always equal and opposite. It is the measure numbers used to write down the vectors that change with the frame.

---

*[drake/systems/plants/RigidBodyTree.h, line 695 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKip_f3yXc_oTOnx6or:-KKip_f3yXc_oTOnx6os:-364626208) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L695)):*
> ```objc
>    margins. If `false`, the representation without margins is used instead.
>    **/
>   void ComputeMaximumDepthCollisionPoints(const KinematicsCache<double> &cache,
> ```

The `&`s here should be next to the type.

---

*[drake/systems/plants/RigidBodyTree.h, line 702 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKipdCUmmJZaU9pKrG4:-KKipdCUmmJZaU9pKrG5:1556261958) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L702)):*
> ```objc
>                                           std::vector<int> &bodyA_idx,
>                                           std::vector<int> &bodyB_idx,
>                                           bool use_margins = true);
> ```

Reorder this from "left to right" so that the first return argument is the "most significant" item (in this case the bodies). That way the documentation for the parameters reads "the bodies that are colliding" then "the points on those bodies" then "the normals at those points" etc. Otherwise you have to use terms you haven't defined yet.

Actually, consider defining and return a vector of structs that can contain all the information in one place. That would be cleaner and would also be extensible without changing the method signature.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-:-KKiqkUK1fObOX4P9fm3:-770342775)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.h, line 671 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKimIx5-syJ7fwUTAYV:-KKjM7iKh0xA-vtbmc8b:-246858960) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L671)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Although this will look lovely in doxygen's documentation, it is hard to read in the header file. It is usually best to compromise so that the text is readable either way, since developers generally only see it in the header. Here I suggest:
> ```
> given `q` and `v`.
> ```
> which looks reasonably good in both places and conveys the idea just as well.
</details>
Suggestion accepted. 

---

*[drake/systems/plants/RigidBodyTree.h, line 677 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKin65RUoZi6sTBzlbC:-KKjMKipHwNP5KiQEtTO:781283130) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L677)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Actually this method doesn't return which collision elements are in contact. Should it? That information would be very difficult to extract from just the body and contact point location. I noticed that your test case actually used the collision elements but that only worked because you knew there was only one on the body.
</details>
That is correct and yes, my test assumes (and for the accompanying sdf that is true) that each body only has one collision element.
Currently Drake does not need to tell what particular elements are in contact. From dynamics, it treats each body as the union of their collision elements. 
What is wrong here is the documentation. It should say `RigidBody`, not `CollisionElement`.

---

*[drake/systems/plants/RigidBodyTree.h, line 682 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKiqT6fHcsSGzqONoh9:-KKjNvi_gCgGHQevLj2z:-1429169295) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L682)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Word this more precisely. The normal vectors are physical quantities independent of the expressed-in frame; they are always equal and opposite. It is the measure numbers used to write down the vectors that change with the frame.
</details>
I just decided to remove this last sentence since, on a second read, I don't think it was providing any extra useful information.

---

*[drake/systems/plants/RigidBodyTree.h, line 695 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKip_f3yXc_oTOnx6or:-KKjOLH43bomCUKfHFVA:676635175) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L695)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> The `&`s here should be next to the type.
</details>
Done

---

*[drake/systems/plants/RigidBodyTree.h, line 702 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKipdCUmmJZaU9pKrG4:-KKjRK3ksE268YZsUcfn:-714113996) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L702)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Reorder this from "left to right" so that the first return argument is the "most significant" item (in this case the bodies). That way the documentation for the parameters reads "the bodies that are colliding" then "the points on those bodies" then "the normals at those points" etc. Otherwise you have to use terms you haven't defined yet.
> 
> Actually, consider defining and return a vector of structs that can contain all the information in one place. That would be cleaner and would also be extensible without changing the method signature.
</details>
I agree with you, but I am using this order to be consistent with the previous API. 
I will try to implement your suggestion.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.h, line 702 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKipdCUmmJZaU9pKrG4:-KKjfh5mn8HY0_q76Ha9:-1458220989) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L702)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> I agree with you, but I am using this order to be consistent with the previous API. 
> I will try to implement your suggestion.
</details>
@sherm1: I still didn't document this properly but, what do you think about this implementation?
Notice there is a [`PointPair`](https://github.com/RobotLocomotion/drake/blob/master/drake/systems/plants/collision/point_pair.h). However that class makes reference to collision elements by id. The struct I am implementing makes a direct reference to the participating rigid bodies by reference. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review pass done.

---

Reviewed 5 of 6 files at r1, 1 of 1 files at r2, 2 of 5 files at r3.
Review status: 5 of 8 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/systems/plants/rigid_body_collision_pair.h, line 4 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjkPZsWtC-ATJ5HQbl:-KKjkPZsWtC-ATJ5HQbm:-1506289950) ([raw file](https://github.com/robotlocomotion/drake/blob/70707cdbadda67e64fb72ed17f679d09237cb80b/drake/systems/plants/rigid_body_collision_pair.h#L4)):*
> ```objc
> 
> #include <Eigen/Dense>
> #include <Eigen/StdVector>
> ```

This include is only for when you want to use `std::vector` with an Eigen vectorizable fixed-size type. You don't have that here so shouldn't include.

---

*[drake/systems/plants/rigid_body_collision_pair.h, line 10 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjiga6mB59YeIOO-wO:-KKjiga6mB59YeIOO-wP:-284031570) ([raw file](https://github.com/robotlocomotion/drake/blob/70707cdbadda67e64fb72ed17f679d09237cb80b/drake/systems/plants/rigid_body_collision_pair.h#L10)):*
> ```objc
> 
> 
> struct DRAKERBM_EXPORT RigidBodyCollisionPair {
> ```

Needs doxygen for the struct. Should remove the "EXPORT" macro and include above since this is header-only. Should have a unit test somewhere (OK to add to an existing test IMO); that's just to make sure no one screws it up in the future.

---

*[drake/systems/plants/rigid_body_collision_pair.h, line 15 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjj1Jx_ql_28OcltX2:-KKjj1Jx_ql_28OcltX3:-2057040057) ([raw file](https://github.com/robotlocomotion/drake/blob/70707cdbadda67e64fb72ed17f679d09237cb80b/drake/systems/plants/rigid_body_collision_pair.h#L15)):*
> ```objc
>                          double distance,
>                          const Eigen::Vector3d ptA,
>                          const Eigen::Vector3d ptB)
> ```

These should be const references `const Eigen::Vector3d& normal` etc.

---

*[drake/systems/plants/rigid_body_collision_pair.h, line 28 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjjH_OS7yh_3jTxtod:-KKjjH_OS7yh_3jTxtoe:-407440265) ([raw file](https://github.com/robotlocomotion/drake/blob/70707cdbadda67e64fb72ed17f679d09237cb80b/drake/systems/plants/rigid_body_collision_pair.h#L28)):*
> ```objc
>   Eigen::Vector3d ptB_;
>   Eigen::Vector3d normal_;
>   double distance_;
> ```

If you intend to expose the data members in a struct, don't use the trailing "_" convention. See style guide [struct vs class](https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes) and [naming rules](https://google.github.io/styleguide/cppguide.html#Variable_Names).

---

*[drake/systems/plants/RigidBodyTree.cpp, line 604 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjlL7n9lOi-DE1iAlC:-KKjlL7n9lOi-DE1iAlD:1621689659) ([raw file](https://github.com/robotlocomotion/drake/blob/55e8d99621b52dce854c07a359bee70f67663c13/drake/systems/plants/RigidBodyTree.cpp#L604)):*
> ```cpp
> 
> void RigidBodyTree::ComputeMaximumDepthCollisionPoints(
>     const KinematicsCache<double> &cache, VectorXd &phi, Matrix3Xd &normal,
> ```

The `&`s here are in the wrong place.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 625 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjlktWdeuhZQlw5uNI:-KKjlktX8xzs0ZKND9Np:-1170686744) ([raw file](https://github.com/robotlocomotion/drake/blob/55e8d99621b52dce854c07a359bee70f67663c13/drake/systems/plants/RigidBodyTree.cpp#L625)):*
> ```cpp
> 
>   for (size_t i = 0; i < num_contact_points; i++) {
>     contact_points[i].getResults(&ptA, &ptB, &n, &distance);
> ```

I don't see the advantage of unpacking these here -- better to return a struct.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 634 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjkvhxchyPGL32YKo1:-KKjkvhxchyPGL32YKo2:562485815) ([raw file](https://github.com/robotlocomotion/drake/blob/55e8d99621b52dce854c07a359bee70f67663c13/drake/systems/plants/RigidBodyTree.cpp#L634)):*
> ```cpp
>     const RigidBody::CollisionElement* elementA =
>         dynamic_cast<const RigidBody::CollisionElement*>(
>             collision_model->readElement(contact_points[i].getIdA()));
> ```

Why does `readElement()` return an `Element` rather than a `CollisionElement`. Having to dynamic cast here seems unnecessarily expensive.

---

*[drake/systems/plants/RigidBodyTree.h, line 677 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKin65RUoZi6sTBzlbC:-KKjfUXf3ZjEoTf3fqE1:-696280472) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L677)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> That is correct and yes, my test assumes (and for the accompanying sdf that is true) that each body only has one collision element.
> Currently Drake does not need to tell what particular elements are in contact. From dynamics, it treats each body as the union of their collision elements. 
> What is wrong here is the documentation. It should say `RigidBody`, not `CollisionElement`.
</details>
This still seems awkward. Say body A has two collision elements and both of those hit a large collision element on body B. I guess we get two collision points with (bodyA,bodyB) returned but can't tell which is which? Is that the behavior you want?

---

*[drake/systems/plants/RigidBodyTree.h, line 702 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKipdCUmmJZaU9pKrG4:-KKjeloBCQeZAc-8X06B:1404828759) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L702)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> @sherm1: I still didn't document this properly but, what do you think about this implementation?
> Notice there is a [`PointPair`](https://github.com/RobotLocomotion/drake/blob/master/drake/systems/plants/collision/point_pair.h). However that class makes reference to collision elements by id. The struct I am implementing makes a direct reference to the participating rigid bodies by reference. 
</details>
Yeah, I see that now. But since you are changing the name of the method I don't think you are in any way stuck with the existing calling sequence -- every call has to be found and changed anyway.

Also, didn't you create a new struct, `RigidBodyCollisionPair` with exactly this info in it? This would be a perfect use for it (or did you add that for this purpose and just haven't installed it yet?)

---

*[drake/systems/plants/RigidBodyTree.h, line 694 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjhe0P4NQHishKVDyu:-KKjhe0P4NQHishKVDyv:-1241311025) ([raw file](https://github.com/robotlocomotion/drake/blob/55e8d99621b52dce854c07a359bee70f67663c13/drake/systems/plants/RigidBodyTree.h#L694)):*
> ```objc
>    margins. If `false`, the representation without margins is used instead.
>    **/
>   void ComputeMaximumDepthCollisionPoints(const KinematicsCache<double>& cache,
> ```

Does this replace some other method that should now be removed?

---

*[drake/systems/plants/test/rigid_body_tree/rbt_collisions_test.cc, line 33 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjm-Bal0F7dimbqu_C:-KKjm-Bal0F7dimbqu_D:-1558306102) ([raw file](https://github.com/robotlocomotion/drake/blob/55e8d99621b52dce854c07a359bee70f67663c13/drake/systems/plants/test/rigid_body_tree/rbt_collisions_test.cc#L33)):*
> ```cpp
> };
> 
> // Solutions are accessed by collision element id using an std::unordered_set.
> ```

comments says "set" but type is "map"

---

*[drake/systems/plants/test/rigid_body_tree/rbt_collisions_test.cc, line 64 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjiG3NVPHjfwuzL_qS:-KKjiG3NVPHjfwuzL_qT:389759718) ([raw file](https://github.com/robotlocomotion/drake/blob/55e8d99621b52dce854c07a359bee70f67663c13/drake/systems/plants/test/rigid_body_tree/rbt_collisions_test.cc#L64)):*
> ```cpp
> };
> 
> TEST_F(RBTCollisionTest, FindAndComputeContactPoints) {
> ```

Should note here that a single maximum-depth collision point is inadequate for supporting a small block on a large block. I presume you are just using this geometry to test the max-depth code; we don't want anyone to look at this test and think that this is the right way to do it!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-:-KKjm4SlJGXMpaC9LI-g:-1270496416)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 5 of 8 files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/systems/plants/rigid_body_collision_pair.h, line 10 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjiga6mB59YeIOO-wO:-KKk7SRIb_pZ9ll0rikP:505254991) ([raw file](https://github.com/robotlocomotion/drake/blob/70707cdbadda67e64fb72ed17f679d09237cb80b/drake/systems/plants/rigid_body_collision_pair.h#L10)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Needs doxygen for the struct. Should remove the "EXPORT" macro and include above since this is header-only. Should have a unit test somewhere (OK to add to an existing test IMO); that's just to make sure no one screws it up in the future.
</details>
This is a struct with all public members. Then a unit test should test the constructor?

---

*[drake/systems/plants/RigidBodyTree.cpp, line 634 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjkvhxchyPGL32YKo1:-KKk4nCcZasuwjWKFUa0:245853349) ([raw file](https://github.com/robotlocomotion/drake/blob/55e8d99621b52dce854c07a359bee70f67663c13/drake/systems/plants/RigidBodyTree.cpp#L634)):*
This is one of the issues fixed by [2299](https://github.com/robotlocomotion/drake/pull/2299), which got reverted. See the last two bullets in the description. I'll add them here for easy reference:

> - Add member RigidBody::collision_elements_. This allows direct access of collision elements in a body instead of having to call Model::readElement (which searches in the model by id).
- A number of dynamics_cast are eliminated from RigidBodyTree.cpp. See here for instance.

Hopefully when we have Gurobi I'll be able to find the issue with Atlas' tests and we'll be able to merge it.

---

*[drake/systems/plants/RigidBodyTree.h, line 677 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKin65RUoZi6sTBzlbC:-KKk5SoUw-gC6gHZhU3y:2010459051) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L677)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> This still seems awkward. Say body A has two collision elements and both of those hit a large collision element on body B. I guess we get two collision points with (bodyA,bodyB) returned but can't tell which is which? Is that the behavior you want?
</details>
That is the behavior we have with the current implementation. Actually, if we had #2299 merged, we could directly return `CollisionElement`'s which purposely have a `RigidBody*`. Not only that, but #2299 adds a `RigidBody::collision_elements_` vector of `CollisionElement*`'s which would allow to obtain collision element's for each point pair if needed. One of the main objectives of #2299 was precisely to facilitate (and speed up) the navigation back and forth between RigidBody's and CollisionElement's.

---

*[drake/systems/plants/RigidBodyTree.h, line 702 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKipdCUmmJZaU9pKrG4:-KKk6VO3pxC96_R1Wos5:102608455) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L702)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Yeah, I see that now. But since you are changing the name of the method I don't think you are in any way stuck with the existing calling sequence -- every call has to be found and changed anyway.
> 
> Also, didn't you create a new struct, `RigidBodyCollisionPair` with exactly this info in it? This would be a perfect use for it (or did you add that for this purpose and just haven't installed it yet?)
</details>
Yes, I did use it. I created it in reply to your comment. I just didn't go ahead and implement gtests and documentation waiting for you to give me a (hopefully positive) feedback before spending time cleaning it up.
I'll just go ahead and do so then.

---

*[drake/systems/plants/RigidBodyTree.h, line 694 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjhe0P4NQHishKVDyu:-KKk6zXvKHVlt2Ex72CA:-343949596) ([raw file](https://github.com/robotlocomotion/drake/blob/55e8d99621b52dce854c07a359bee70f67663c13/drake/systems/plants/RigidBodyTree.h#L694)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Does this replace some other method that should now be removed?
</details>
No, it is a new method. The plan is to use it within `RigidBodySystem::dynamics`. In a next PR when I submit the Prius demo

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
A few responses.

---

Reviewed 3 of 5 files at r3.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/systems/plants/rigid_body_collision_pair.h, line 10 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjiga6mB59YeIOO-wO:-KKkBjqVqTnRdjgCz6Kd:-844365266) ([raw file](https://github.com/robotlocomotion/drake/blob/70707cdbadda67e64fb72ed17f679d09237cb80b/drake/systems/plants/rigid_body_collision_pair.h#L10)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> This is a struct with all public members. Then a unit test should test the constructor?
</details>
Well, it appears to be part of the public API so it ought to have a unit test, which would be very simple! If you object to that it would be fine with me as long as some other unit tests is making use of this class in a way that would fail if this class was broken. However, I didn't see anything like that.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 634 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjkvhxchyPGL32YKo1:-KKkCVJg7sUjtH0HTg29:1079838874) ([raw file](https://github.com/robotlocomotion/drake/blob/55e8d99621b52dce854c07a359bee70f67663c13/drake/systems/plants/RigidBodyTree.cpp#L634)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> This is one of the issues fixed by [2299](https://github.com/robotlocomotion/drake/pull/2299), which got reverted. See the last two bullets in the description. I'll add them here for easy reference:
> 
> > - Add member RigidBody::collision_elements_. This allows direct access of collision elements in a body instead of having to call Model::readElement (which searches in the model by id).
> - A number of dynamics_cast are eliminated from RigidBodyTree.cpp. See here for instance.
> 
> Hopefully when we have Gurobi I'll be able to find the issue with Atlas' tests and we'll be able to merge it.
</details>
Oh yeah, I remember. Would it make sense to make that change with this PR instead? If so that would be a way to reduce the size of the big PR that ran into trouble. One possibility would be simply to move the dynamic_cast into `readElement()` (rename to `FindCollisionElement()?`) with a TODO saying to get rid of it. Then the dynamic_casts wouldn't have to be sprinkled around whenever `FindCollisionElement()` gets called.

---

*[drake/systems/plants/RigidBodyTree.h, line 677 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKin65RUoZi6sTBzlbC:-KKkDlkoLtGGO3jWB1rd:-1234268303) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L677)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> That is the behavior we have with the current implementation. Actually, if we had #2299 merged, we could directly return `CollisionElement`'s which purposely have a `RigidBody*`. Not only that, but #2299 adds a `RigidBody::collision_elements_` vector of `CollisionElement*`'s which would allow to obtain collision element's for each point pair if needed. One of the main objectives of #2299 was precisely to facilitate (and speed up) the navigation back and forth between RigidBody's and CollisionElement's.
</details>
Sigh. Yeah, we need #2299! What about fixing the API in this PR but not the underlying implementation? That would help reduce the size of 2299 (or its successor) when it comes back.

---

*[drake/systems/plants/RigidBodyTree.h, line 702 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKipdCUmmJZaU9pKrG4:-KKkEGyo6182f-25-hSa:2002054205) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L702)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Yes, I did use it. I created it in reply to your comment. I just didn't go ahead and implement gtests and documentation waiting for you to give me a (hopefully positive) feedback before spending time cleaning it up.
> I'll just go ahead and do so then.
</details>
OK, sweet -- I see you are using return value optimization!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-:-KKkF0xmHPrcq1cK-dkH:645610600)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 634 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjkvhxchyPGL32YKo1:-KKoJ00fy6l-iyhp4Q_-:-680566493) ([raw file](https://github.com/robotlocomotion/drake/blob/55e8d99621b52dce854c07a359bee70f67663c13/drake/systems/plants/RigidBodyTree.cpp#L634)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Oh yeah, I remember. Would it make sense to make that change with this PR instead? If so that would be a way to reduce the size of the big PR that ran into trouble. One possibility would be simply to move the dynamic_cast into `readElement()` (rename to `FindCollisionElement()?`) with a TODO saying to get rid of it. Then the dynamic_casts wouldn't have to be sprinkled around whenever `FindCollisionElement()` gets called.
</details>
@sherm1: I pushed #2607 to solve these issues. PTAL

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 0 of 6 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 604 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjlL7n9lOi-DE1iAlC:-KLIjA5clNS0H1H7LB2Y:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/55e8d99621b52dce854c07a359bee70f67663c13/drake/systems/plants/RigidBodyTree.cpp#L604)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> The `&`s here are in the wrong place.
</details>
Done.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 625 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjlktWdeuhZQlw5uNI:-KLIjJkkrHL6eelh3B25:565703351) ([raw file](https://github.com/robotlocomotion/drake/blob/55e8d99621b52dce854c07a359bee70f67663c13/drake/systems/plants/RigidBodyTree.cpp#L625)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> I don't see the advantage of unpacking these here -- better to return a struct.
</details>
#2607 paid off very nicely here. See how now this got simplified quite a bit. Removed the extra unpacking also.

---

*[drake/systems/plants/RigidBodyTree.h, line 677 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKin65RUoZi6sTBzlbC:-KLIjw5oeRBk0mycvFyk:1042983904) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L677)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Sigh. Yeah, we need #2299! What about fixing the API in this PR but not the underlying implementation? That would help reduce the size of 2299 (or its successor) when it comes back.
</details>
The new `PointPair` returned contains direct pointers to the participating collision elements (a per #2607). If the body is needed the user can always call `get_body()` on the collision element now.

---

*[drake/systems/plants/test/rigid_body_tree/rbt_collisions_test.cc, line 33 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjm-Bal0F7dimbqu_C:-KLI7VeJvevz8D_XF3cP:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/55e8d99621b52dce854c07a359bee70f67663c13/drake/systems/plants/test/rigid_body_tree/rbt_collisions_test.cc#L33)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> comments says "set" but type is "map"
</details>
Done.

---

*[drake/systems/plants/test/rigid_body_tree/rbt_collisions_test.cc, line 64 \[r2\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjiG3NVPHjfwuzL_qS:-KLIDPWzlhsn6fSPWgTW:-1434888137) ([raw file](https://github.com/robotlocomotion/drake/blob/55e8d99621b52dce854c07a359bee70f67663c13/drake/systems/plants/test/rigid_body_tree/rbt_collisions_test.cc#L64)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Should note here that a single maximum-depth collision point is inadequate for supporting a small block on a large block. I presume you are just using this geometry to test the max-depth code; we don't want anyone to look at this test and think that this is the right way to do it!
</details>
To avoid confusion I changed this test to be between a sphere and a box, with a single collision point. I also added some documentation at the top.

---

*[drake/systems/plants/rigid_body_collision_pair.h, line 4 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjkPZsWtC-ATJ5HQbl:-KLIiJv8Nx4H0obgAK9G:1592083558) ([raw file](https://github.com/robotlocomotion/drake/blob/70707cdbadda67e64fb72ed17f679d09237cb80b/drake/systems/plants/rigid_body_collision_pair.h#L4)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> This include is only for when you want to use `std::vector` with an Eigen vectorizable fixed-size type. You don't have that here so shouldn't include.
</details>
#2607 started to pay off. I removed this class completely since `PointPair` can now get the job done. I decided to make `PointPair` a struct (will remove public interfaces soon).

---

*[drake/systems/plants/rigid_body_collision_pair.h, line 10 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjiga6mB59YeIOO-wO:-KLIifOGAKQiNGMapHEC:1787864951) ([raw file](https://github.com/robotlocomotion/drake/blob/70707cdbadda67e64fb72ed17f679d09237cb80b/drake/systems/plants/rigid_body_collision_pair.h#L10)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Well, it appears to be part of the public API so it ought to have a unit test, which would be very simple! If you object to that it would be fine with me as long as some other unit tests is making use of this class in a way that would fail if this class was broken. However, I didn't see anything like that.
</details>
This class is now removed and `PointPair` is used. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 3 of 10 files at r4.
Review status: 2 of 13 files reviewed at latest revision, 11 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 610 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLJ-fZs4ytTTETAdy9j:-KLJ-fZs4ytTTETAdy9k:1550123897) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfc11df874dc0eb5e0583f66fcadd311e432ea8/drake/systems/plants/RigidBodyTree.cpp#L610)):*
> ```cpp
>     const RigidBody& bodyA = *contact_points[i].elementA_->get_body();
>     Isometry3d TA;
>     if (bodyA.hasParent()) {  // body is dynamic.
> ```

This is inconsistent with your use of "dynamic" and "static" elsewhere, because you said that if a body is fixed to world then that body is static. Here it would be considered dynamic since it has a parent. Please fix comment.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 613 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLJ054NSHYs6CHBb_IS:-KLJ054NSHYs6CHBb_IT:1074604577) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfc11df874dc0eb5e0583f66fcadd311e432ea8/drake/systems/plants/RigidBodyTree.cpp#L613)):*
> ```cpp
>       TA = cache.getElement(bodyA).transform_to_world;
>     } else {  // body is static.
>       TA = Isometry3d::Identity();
> ```

BTW, for future note it is better to have a "world" entry in the cache that contains an identity transform so that code everywhere else doesn't have to be special-cased.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 618 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLJ-xo0Nnl1vK8lqZuq:-KLJ-xo0Nnl1vK8lqZur:-36924186) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfc11df874dc0eb5e0583f66fcadd311e432ea8/drake/systems/plants/RigidBodyTree.cpp#L618)):*
> ```cpp
>     const RigidBody& bodyB = *contact_points[i].elementB_->get_body();
>     Isometry3d TB;
>     if (bodyB.hasParent()) {  // body is dynamic.
> ```

same comments as above

---

*[drake/systems/plants/RigidBodyTree.h, line 556 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLJ0gYTU2OTSH-J-IPe:-KLJ0gYTU2OTSH-J-IPf:678074825) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfc11df874dc0eb5e0583f66fcadd311e432ea8/drake/systems/plants/RigidBodyTree.h#L556)):*
> ```objc
>    given `q` and `v`.
> 
>    Collision points are returned as a vector of PontPair's.
> ```

typo

---

*[drake/systems/plants/test/rigid_body_tree/rbt_collisions_test.cc, line 71 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLJ1gn1crSbM8md2TLv:-KLJ1gn1crSbM8md2TLw:-1381946327) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfc11df874dc0eb5e0583f66fcadd311e432ea8/drake/systems/plants/test/rigid_body_tree/rbt_collisions_test.cc#L71)):*
> ```cpp
> // RigidBodyTree::ComputeMaximumDepthCollisionPoints.
> // The test consists on finding the maximum depth penetration point between a
> // sphere and a box. The collision point is reported in the frame of the body.
> ```

in the frame of which body?

---

*[drake/systems/plants/test/rigid_body_tree/rbt_collisions_test.cc, line 76 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLJ1l-QWy30RxBJStn5:-KLJ1l-QWy30RxBJStn6:-294141872) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfc11df874dc0eb5e0583f66fcadd311e432ea8/drake/systems/plants/test/rigid_body_tree/rbt_collisions_test.cc#L76)):*
> ```cpp
>   // Its magnitude was chosen to be the minimum value for which these tests can
>   // successfully pass.
>   tolerance_ = 4.0*Eigen::NumTraits<double>::epsilon();
> ```

BTW, nice!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review pass done.

---

Reviewed 4 of 10 files at r4, 9 of 9 files at r5.
Review status: 9 of 13 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.h, line 677 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKin65RUoZi6sTBzlbC:-KLJ5vxdcLYbBiFk4YBG:-823943206) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L677)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> The new `PointPair` returned contains direct pointers to the participating collision elements (a per #2607). If the body is needed the user can always call `get_body()` on the collision element now.
</details>
I can't see that -- it looks like PointPair has Element pointers but no body pointers. Am I misreading something?

---

*[drake/systems/plants/collision/bullet_model.cc, line 482 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLJ3QmD7geIztpLanVO:-KLJ3QmD7geIztpLanVP:1656001064) ([raw file](https://github.com/robotlocomotion/drake/blob/d63ab5b435bf7276d02e80be43fd92a90b999642/drake/systems/plants/collision/bullet_model.cc#L482)):*
> ```cpp
>   auto bt_objA_iter = bt_world.bt_collision_objects.find(idA);
>   if (bt_objA_iter == bt_world.bt_collision_objects.end())
>     throw std::runtime_error(
> ```

I don't think this is a `runtime_error` (unless it came out of a file or something else you couldn't know until runtime). Could be a `logic_error` if it is a user programming mistake; otherwise, it should be a `DRAKE_ASSERT()` or better, `DRAKE_ABORT_UNLESS()`.

---

*[drake/systems/plants/collision/point_pair.h, line 49 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLJ4CmIdvpVAvtbgOb9:-KLJ4CmIdvpVAvtbgObA:-320219675) ([raw file](https://github.com/robotlocomotion/drake/blob/d63ab5b435bf7276d02e80be43fd92a90b999642/drake/systems/plants/collision/point_pair.h#L49)):*
> ```objc
> 
>   /** Element A in the pair participating in the collision. **/
>   const Element* elementA_{nullptr};
> ```

Per styleguide, structs don't use the trailing `_` convention. Consider making this a a class with accessors. It already has too much functionality for a struct according to the styleguide [here](https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-:-KLJ64bT9Fqzer6z7q22:-1270496416)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
@sherm1: Notice that I removed the struct `RigidBodyCollisionPairs` (introduced in this PR by myself) and instead updated `PointPair` to be a very simple struct used to return collision results as you requested for `ComputeMaximumDepthCollisionPoints`. 
While doing this I also completely removed the `ResultCollector` class which turned out to be over complicating things and making unnecessary copies to return results. PTAL

---

Review status: 9 of 13 files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 610 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLJ-fZs4ytTTETAdy9j:-KLJ5dpnhFc3nPt6wDw1:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfc11df874dc0eb5e0583f66fcadd311e432ea8/drake/systems/plants/RigidBodyTree.cpp#L610)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> This is inconsistent with your use of "dynamic" and "static" elsewhere, because you said that if a body is fixed to world then that body is static. Here it would be considered dynamic since it has a parent. Please fix comment.
</details>
Done.

---

*[drake/systems/plants/RigidBodyTree.cpp, line 613 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLJ054NSHYs6CHBb_IS:-KLJ6Oem8_s_1utl4CFh:-1385586327) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfc11df874dc0eb5e0583f66fcadd311e432ea8/drake/systems/plants/RigidBodyTree.cpp#L613)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, for future note it is better to have a "world" entry in the cache that contains an identity transform so that code everywhere else doesn't have to be special-cased.
</details>
I agree but I don't really know how long it would take to implement this in Drake. I created issue #2665 to track this.

---

*[drake/systems/plants/RigidBodyTree.h, line 677 \[r1\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKin65RUoZi6sTBzlbC:-KLJ6eyP1oFCvU46iy_J:-572633423) ([raw file](https://github.com/robotlocomotion/drake/blob/a3be73544128a3a95e2144c2e31fe3a0d5a1236c/drake/systems/plants/RigidBodyTree.h#L677)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> I can't see that -- it looks like PointPair has Element pointers but no body pointers. Am I misreading something?
</details>
The idea is that you do `point_pair.elementA->get_body()`.

---

*[drake/systems/plants/RigidBodyTree.h, line 556 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLJ0gYTU2OTSH-J-IPe:-KLJAomBbHJHGntZVAyr:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfc11df874dc0eb5e0583f66fcadd311e432ea8/drake/systems/plants/RigidBodyTree.h#L556)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> typo
</details>
Done.

---

*[drake/systems/plants/collision/bullet_model.cc, line 482 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLJ3QmD7geIztpLanVO:-KLJ9634KgwhUDfPEORU:1102870210) ([raw file](https://github.com/robotlocomotion/drake/blob/d63ab5b435bf7276d02e80be43fd92a90b999642/drake/systems/plants/collision/bullet_model.cc#L482)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> I don't think this is a `runtime_error` (unless it came out of a file or something else you couldn't know until runtime). Could be a `logic_error` if it is a user programming mistake; otherwise, it should be a `DRAKE_ASSERT()` or better, `DRAKE_ABORT_UNLESS()`.
</details>
Done. What's the difference between DRAKE_ASSERT and DRAKE_ASSERT_UNLESS?

---

*[drake/systems/plants/collision/point_pair.h, line 49 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLJ4CmIdvpVAvtbgOb9:-KLJ9If2hYYWN-g3Qj6N:1982324553) ([raw file](https://github.com/robotlocomotion/drake/blob/d63ab5b435bf7276d02e80be43fd92a90b999642/drake/systems/plants/collision/point_pair.h#L49)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> Per styleguide, structs don't use the trailing `_` convention. Consider making this a a class with accessors. It already has too much functionality for a struct according to the styleguide [here](https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes).
</details>
I refactored this class to be a simple struct. Now variables do not have trailing `_`'s

---

*[drake/systems/plants/test/rigid_body_tree/rbt_collisions_test.cc, line 71 \[r4\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLJ1gn1crSbM8md2TLv:-KLJAEybMfdwLoTiqczC:-706905406) ([raw file](https://github.com/robotlocomotion/drake/blob/bbfc11df874dc0eb5e0583f66fcadd311e432ea8/drake/systems/plants/test/rigid_body_tree/rbt_collisions_test.cc#L71)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> in the frame of which body?
</details>
I realize that comment does not belong here. The answer to your question however is in the doc for `ComputeMaximumDepthCollisionPoints`.

---

*[drake/systems/plants/rigid_body_collision_pair.h, line 15 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjj1Jx_ql_28OcltX2:-KLJAb2ZGKXar8Jpzb1Y:-499146009) ([raw file](https://github.com/robotlocomotion/drake/blob/70707cdbadda67e64fb72ed17f679d09237cb80b/drake/systems/plants/rigid_body_collision_pair.h#L15)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> These should be const references `const Eigen::Vector3d& normal` etc.
</details>
Done.

---

*[drake/systems/plants/rigid_body_collision_pair.h, line 28 \[r3\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KKjjH_OS7yh_3jTxtod:-KLJAbLw59dSNYp5aRmX:-1104670426) ([raw file](https://github.com/robotlocomotion/drake/blob/70707cdbadda67e64fb72ed17f679d09237cb80b/drake/systems/plants/rigid_body_collision_pair.h#L28)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> If you intend to expose the data members in a struct, don't use the trailing "_" convention. See style guide [struct vs class](https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes) and [naming rules](https://google.github.io/styleguide/cppguide.html#Variable_Names).
</details>
Updated `PointPair` struct does not use the trailing "_"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-:-KLJAquSkeiEy7zrrgcE:-1306837868)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 4 files at r6, 5 of 5 files at r7.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/systems/plants/collision/bullet_model.cc, line 482 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLJ3QmD7geIztpLanVO:-KLJTKqVlpPymUudg2Nu:753796137) ([raw file](https://github.com/robotlocomotion/drake/blob/d63ab5b435bf7276d02e80be43fd92a90b999642/drake/systems/plants/collision/bullet_model.cc#L482)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary>
> Done. What's the difference between DRAKE_ASSERT and DRAKE_ASSERT_UNLESS?
</details>
LGTM, but consider using `DRAKE_ABORT_UNLESS`. That is the same as `DRAKE_ASSERT` except that it remains in place in Release builds. `DRAKE_ASSERT` is only present in Debug which makes sense for assertions that could have a noticeable performance cost. If you are doing something must more expensive than the test then it is better to keep the assertions in place all the time.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-:-KLJUj6SSPnlExXYKfaO:1417088721)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/systems/plants/test/rigid_body_tree/small_sphere_on_large_box.sdf, line 28 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLLVxjIRoy4DV9kn5bL:-KLLVxjIRoy4DV9kn5bM:-968933547) ([raw file](https://github.com/robotlocomotion/drake/blob/59a76845285e848feee714cb32cb758a35628e29/drake/systems/plants/test/rigid_body_tree/small_sphere_on_large_box.sdf#L28)):*
> ```
>         <geometry>
>           <sphere>
> 	    <radius>0.5</radius>
> ```

BTW, are there some tabs in this file? If so, please convert to spaces!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[drake/systems/plants/collision/bullet_model.cc, line 482 \[r5\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLJ3QmD7geIztpLanVO:-KLLs5Ccz9_j76ik8Q75:-1474730981) ([raw file](https://github.com/robotlocomotion/drake/blob/d63ab5b435bf7276d02e80be43fd92a90b999642/drake/systems/plants/collision/bullet_model.cc#L482)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> LGTM, but consider using `DRAKE_ABORT_UNLESS`. That is the same as `DRAKE_ASSERT` except that it remains in place in Release builds. `DRAKE_ASSERT` is only present in Debug which makes sense for assertions that could have a noticeable performance cost. If you are doing something must more expensive than the test then it is better to keep the assertions in place all the time.
</details>
Oh, I see. I was misreading  `DRAKE_ASSERT_UNLESS` which of course didn't make any sense. I considered this change since in any case by #2662 pointers to elements will be immediately available (no find needed).

---

*[drake/systems/plants/test/rigid_body_tree/small_sphere_on_large_box.sdf, line 28 \[r7\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLLVxjIRoy4DV9kn5bL:-KLLsZTrGYP_W2qE8V3S:-853947946) ([raw file](https://github.com/robotlocomotion/drake/blob/59a76845285e848feee714cb32cb758a35628e29/drake/systems/plants/test/rigid_body_tree/small_sphere_on_large_box.sdf#L28)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> BTW, are there some tabs in this file? If so, please convert to spaces!
</details>
nice catch! I didn't had my emacs set up to work with c++ but not with xml files!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
@liangfok, @sherm1: Besides two unresolved discussions I think this is ready to merge?

---

Review status: 10 of 13 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-:-KLN03nJaky_DN0yhTCn:-374575643)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Reviewed 1 of 10 files at r4, 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/plants/collision/bullet_model.cc, line 489 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLN9Lj6hHYSjPifbYIz:-KLN9Lj6hHYSjPifbYJ-:2014047482) ([raw file](https://github.com/robotlocomotion/drake/blob/c8c201111eb24ac3ed9baa75d2ac879422dd0067/drake/systems/plants/collision/bullet_model.cc#L489)):*
> ```cpp
>   DRAKE_ABORT_UNLESS(bt_objB_iter != bt_world.bt_collision_objects.end() &&
>         "In BulletModel::findClosestPointsBetweenElements: "
>             "invalid ElementId for body B.");
> ```

BTW, @jwnimmer-tri check this out. First use I've seen of `DRAKE_ABORT_UNLESS` in the wild (by @amcastro-tri).  Would be nicer with a message argument.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-:-KLNAfrnebF4L5A6JM28:1417088721)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[drake/systems/plants/collision/bullet_model.cc, line 489 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLN9Lj6hHYSjPifbYIz:-KLNBHwnQDlVD3ikUAJC:-435572027) ([raw file](https://github.com/robotlocomotion/drake/blob/c8c201111eb24ac3ed9baa75d2ac879422dd0067/drake/systems/plants/collision/bullet_model.cc#L489)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary>
> BTW, @jwnimmer-tri check this out. First use I've seen of `DRAKE_ABORT_UNLESS` in the wild (by @amcastro-tri).  Would be nicer with a message argument.
</details>
You guys can add message variants if you want.  Personally, I think they are an anti-pattern.  The file name is already in the message (so saying "BulletModel" here is redundant), the function name is already in the message (thus, the "findClosest..." is redundant!), the file + line number is way more precise that any message, they should trip rarely enough that the additional text is of low value, the extra message text obscures the readability of the non-error-handling portions of the code, and the message is never unit tested, and thus probably wrong (as demonstrated here) because it's an abort, not an exception.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
*[drake/systems/plants/collision/bullet_model.cc, line 489 \[r8\]](https://reviewable.io:443/reviews/robotlocomotion/drake/2596#-KLN9Lj6hHYSjPifbYIz:-KLNH-x_l3AMriMPuE2l:1210766583) ([raw file](https://github.com/robotlocomotion/drake/blob/c8c201111eb24ac3ed9baa75d2ac879422dd0067/drake/systems/plants/collision/bullet_model.cc#L489)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> You guys can add message variants if you want.  Personally, I think they are an anti-pattern.  The file name is already in the message (so saying "BulletModel" here is redundant), the function name is already in the message (thus, the "findClosest..." is redundant!), the file + line number is way more precise that any message, they should trip rarely enough that the additional text is of low value, the extra message text obscures the readability of the non-error-handling portions of the code, and the message is never unit tested, and thus probably wrong (as demonstrated here) because it's an abort, not an exception.
</details>
Those are very good points @jwnimmer-tri! @amcastro-tri, now I agree with Jeremy that you shouldn't attempt to add a message to these. If you really want to, you should shrink it down to the part that carries new information.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/2596)*
<!-- Sent from Reviewable.io -->
